packages/menu/src/Menu.tsx

```diff
@@ -88,46 +89,152 @@ function Menu({
+          throw new Error('SubMenu titles must be unique');
+        }
+
+        titleArr.push(child.props.title);
+
+        if (child.props.active && !hasSetInitialOpen.current) {
+          setCurrentSubMenu(child);
+          hasSetInitialOpen.current = true;
+        }
+
+        const isCurrentSubMenu =
+          currentSubMenu?.props.title === child?.props.title;
+
+        return React.cloneElement(child, {
+          ref: setRef,
+          open: isCurrentSubMenu,
+          setOpen: (state: boolean) => {
+            if (currentChildRef) {
+              setFocused(currentChildRef);
+            }
+
+            setCurrentSubMenu(state ? child : null);
+          },
+          onKeyDown: (e: KeyboardEvent) => {
+            if (e.keyCode === keyMap.ArrowLeft && isCurrentSubMenu) {
+              setCurrentSubMenu(null);
+            }
+
+            if (e.keyCode === keyMap.ArrowRight) {
+              setCurrentSubMenu(child);
+            }
+          },
+          onFocus,
+          children: updateChildren(child.props.children),
+          onExited: () => {
+            setClosed(curr => !curr);
+          },
+        });
+      }
+
+      if (
+        isComponentType<'MenuItem'>(child, 'MenuItem') ||
+        isComponentType<'FocusableMenuItem'>(child, 'FocusableMenuItem')
+      ) {
+        return React.cloneElement(child, {
+          ref: setRef,
+          onFocus,
+        });
+      }
+
+      if (child.props?.children) {
+        return React.cloneElement(child, {
+          children: updateChildren(child.props.children),
+        });
+      }
+
+      return child;
+    });
+  }
 
   const [focused, setFocused] = useState<HTMLElement>(refs[0] || null);
-  const isControlled = typeof controlledOpen === 'boolean';
   const [uncontrolledOpen, uncontrolledSetOpen] = useState(false);
-  const open = isControlled ? controlledOpen : uncontrolledOpen;
-  const setOpen = isControlled ? controlledSetOpen : uncontrolledSetOpen;
+  const setOpen =
+    (typeof controlledOpen === 'boolean' && controlledSetOpen) ||
+    uncontrolledSetOpen;
+  const open = controlledOpen ?? uncontrolledOpen;
+
+  // When a SubMenu becomes open, it's set to currentSubMenu, and we focus on the first child.
+  useEffect(() => {
+    if (!currentSubMenu) {
+      const focusedRefIndex = refs.indexOf(focused);
+      const subMenu = refs[focusedRefIndex];
+
+      subMenu?.focus();
+      setUsingKeyboard(true);
+    } else {
+      const focusedRefIndex = refs.indexOf(focused);
+      const subMenuFirstChild = refs[focusedRefIndex + 1];
+
+      subMenuFirstChild?.focus();
+      setUsingKeyboard(true);
+    }
+  }, [currentSubMenu]);
+
+  const updatedChildren = React.useMemo(() => updateChildren(children), [
+    children,
+    focused,
+    currentSubMenu,
+    open,
+    refs,
+  ]);
 
   const popoverRef: React.RefObject<HTMLUListElement> = useRef(null);
 
+  const setFocus = (el: HTMLElement) => {
+    setFocused(el);
+    el?.focus();
```

Comment:
Rather than checking here if `el` exists, we should probably make the function return on the first line if el is undefined. Otherwise we may have a bug in the future where we pass `undefined` to `setFocused()`.

Reply:
👍